### ISSUE TITLE
HDR not being output on an Apple Pro XDR display ?

### ISSUE THREAD

**Issue by swa66** [iina/iina#3236]
I tried many sources, never got any of them to properly output HDR onto an Apple Pro XDR Display.
The MP7,1 and the XDR display do output HDR (in chrome, firefox, Quicktime, etc.)

Mostly the range of colors IINA produces is "compressed" into the normal range and not "extended" to include the high brightness the monitors are capable of. Resulting is a dark to very dark result instead of a normal result very very bright highlights.

It's not just one source, but I'll just use this one as it's easy to play this in multiple browsers and other media players, so we can easily see other players can play this content.

Youtube: https://www.youtube.com/watch?v=2J7xlDH4QkA
Plays as expected in Chrome and Firefox and Safari
You can download the file before Google got their hands on it as well: 
https://mega.nz/file/qdVg0IIK#B0L811GNQkvnEpZsE5558cpTKCaEGzohHwRjF6-3_gs
And that file plays as expected in Quicktime

Neither the youtube URL nor the downloaded file is output as HDR content on the same machine/display/settings/.... neither in a window nor in full screen.

These demos use significant HDR effects to showcase it, so it's easy to spot when it's missing.

IINA does parse the files, and shows them, it's just not outputting it as HDR.

Obviously for files such as these there's easy workarounds, but that's not always the case.

**Comment by kkanungo17** [comment]
This is probably because mpv doesn't support passing through HDR metadata to the display, it instead calculates the brightness values itself.

You can try messing with the HDR options in mpv to see if you can get a similar output.

**Comment by varenc** [comment]
Please only file issues on IINA if they are unique to IINA, and not just issues with the underlying MPV player.  Try using MPV directly and see if you can reproduce the problem.  I suspect it's the same in MPV, meaning that's where you should file the issue.  This one looks related https://github.com/mpv-player/mpv/issues/7341

The template for a new issue should include "- [ ] MPV does not have this problem." to cover this situation.

**Comment by Akemi** [comment]
that's not completely right. this is a problem not solvable alone by mpv with the way HDR works on macOS, i believe. 

to get HDR the right colour space has to be set on the layer itself, so the tonemapping and related mechanisms of the OS can do its work. libmpv doesn't know about the layer itself and this has to be done by the libmpv user itself. so even if the mentioned mpv issue is resolved that wouldn't help iina, besides maybe as code reference.

**Comment by anastasiuspernat** [comment]
Yeah, @Akemi is correct! Here's a way to enable HDR on mpv using the following hack: https://github.com/mpv-player/mpv/pull/8387
It does work in mpv but it doesn't work in IINA with libmpv. As @Akemi noticed to enable this in IINA the `wantsExtendedDynamicRangeContent` has to be set to `true` on the layer and also `wantsExtendedDynamicRangeOpenGLSurface` set to `true` on the `NSWindow` itself.

**Comment by anastasiuspernat** [comment]
So I tried quick experiments, setting `wantsExtendedDynamicRangeContent` to `true` in `ViewLayer.swift` and also `wantsExtendedDynamicRangeOpenGLSurface` in `VideoView.swift` (using HDR-enabled libmpv build). It didn't turn on HDR mode on the video. Probably something else needs to be sorted out as well.

**Comment by anastasiuspernat** [comment]
So I figured it out, mostly! Works on my end for all videos.

But I need support from IINA developers to make it work in all cases. 
There are two places that need to be changed: 
1) `ViewLayer.swift` add
~~~
    wantsExtendedDynamicRangeContent = true
    // Activates HDR color space
    var name = CGColorSpace.displayP3_PQ_EOTF
    colorspace = CGColorSpace(name: name)
~~~ 
The color space's `name` must be taken from Video Info's `Mastering display color primaries` combined with `Transfer characteristics`.
This is enough to activate and test HDR. But it's not enough to produce correct HDR image.

2) There's a `target-trc` option in `mpv`. This one should be set to `Transfer characteristics` of the Video Info. I couldn't find a way to set it through IINA. Using advanced options doesn't help.

**Comment by anastasiuspernat** [comment]
So if anybody is still interested in HDR/XDR monitor support, there's a way to embody it with auto-detection of the color space (`Mastering display color primaries` & `Transfer characteristics`) on IINA side. 
Test code works perfectly on my end on LG 27GN950-B monitor (DisplayHDR 600). HDR footage is displayed correctly, the same way QuickTime displays it natively. Tried many different videos with various codecs and gamma curves. Should work on any other monitor supported by macOS Catalina+ (Apple Pro XDR included).

I don't want to sound like a secret keeper it's just there's not enough hours in the day time (I run a CG software business 24/7). Can try to finalize and make pull request, but not sure about the stage of IINA contribution. I tried to make a simple (unconnected) pull request that is still hanging around.

**Comment by varenc** [comment]
Yea I'm not so sure about the state of PRs either... but this is awesome enough that I think it'd be worth posting it at least so others like I can try it out! And it might make its way into master eventually.  Or just drop a .patch on this issue if that's easier.

(and sorry for initially assuming this was an MPV bug.  I've since played with some HDR media and have noticed this issue too! I also noticed that IINA has different `video-out-params/sig-peak` values from MPV. 🤷‍♂️)

**Comment by EkilDeew** [comment]
Is this only supported for XDR display? Tried it on the 16" display and hdr does indeed work but can't find a way to get proper colors.

**Comment by axfelix** [comment]
Is this any closer to being merged? @alejx this seems like the most complete proposed fix I've seen to IINA's HDR
